mesonbuild/dependencies/blas_lapack.py

```diff
@@ -0,0 +1,136 @@
+def openblas_factory(env: 'Environment', for_machine: 'MachineChoice',
+                  kwargs: T.Dict[str, T.Any],
+                  methods: T.List[DependencyMethods]) -> T.List['DependencyGenerator']:
+    candidates: T.List['DependencyGenerator'] = []
+
+    if DependencyMethods.CMAKE in methods:
+        candidates.append(functools.partial(
+            CMakeDependency, 'OpenBLAS', env, kwargs))
+
+    if DependencyMethods.SYSTEM in methods:
+        candidates.append(functools.partial(OpenBLASDependencySystem, 'openblas', env, kwargs))
+
+    return candidates
+
+
+class OpenBLASDependencySystem(SystemDependency):
+    def __init__(self, name: str, environment: 'Environment', kwargs: T.Dict[str, T.Any]) -> None:
+        super().__init__(name, environment, kwargs)
+        self.feature_since = ('0.64.0', '')
+
+        self.env = environment
+        self.name = 'openblas'
+        self.is_found = False
+
+        # First, look for paths specified in a machine file
+        props = self.env.properties[self.for_machine].properties
+        if any(x in props for x in ['openblas_includedir', 'openblas_librarydir']):
+            self.detect_openblas_machine_file(props)
+
+        # Then look in standard directories by attempting to link
+        if not self.is_found:
+            extra_dirs = []  # TODO - for Windows may be, e.g.: Path('C:/opt/openblas/if_32/64/')
+            link_arg = self.clib_compiler.find_library('openblas', self.env, extra_dirs)
+            h = self.clib_compiler.has_header('openblas_config.h', '', self.env, dependencies=[self])
+            if link_arg and h[0]:
+                self.is_found = True
+                self.link_args += link_arg
+
+        if self.is_found:
+            self.version = self.detect_openblas_version()
+
+        self.run_check()
+
+    def detect_openblas_machine_file(self, props: dict) -> None:
+        incdir = props.get('openblas_includedir')
+        assert incdir is None or isinstance(incdir, str)
+        libdir = props.get('openblas_librarydir')
+        assert libdir is None or isinstance(libdir, str)
+
+        if incdir and libdir:
+            self.is_found = True
+            inc_dir = Path(incdir)
+            lib_dir = Path(libdir)
+            if not inc_dir.is_absolute() or not lib_dir.is_absolute():
+                # TODO: why doesn't this fail the build? It gets caught, how to avoid?
+                raise DependencyException('Paths given for openblas_includedir and openblas_librarydir in machine file must be absolute')
+        elif incdir or libdir:
+            raise DependencyException('Both openblas_includedir *and* openblas_librarydir have to be set in your machine file (one is not enough)')
+        else:
+            # We only call this method if incdir or libdir were found
+            raise RuntimeError('Meson internal issue during openblas dependency detection')
```

Comment:
```
raise mesonlib.MesonBugException('XXXXXXX')
```

yields the error:
```
meson.build:lineno:colno: ERROR: XXXXX

This is a Meson bug and should be reported!
```

It is better than RuntimeError because it explicitly denotes an internal issue.